### PR TITLE
CFE-3091: Added debug_abort_if_reached macro

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

File: libutils/condition_macros.h [hunk 1]
```diff
@@ -3,10 +3,19 @@
 
 #include <assert.h>
 
+// This function contains macros to use for error checking, assertions,
```
Comment (vpodzime) [inline libutils/condition_macros.h]: *This file...*

File: libutils/condition_macros.h [hunk 2]
```diff
@@ -17,4 +26,13 @@
 }
 #endif
 
+// Similar to assert_or_return, except you put it inside the if
+// body which handles the error in release builds
+#ifndef debug_abort_if_reached
+#define debug_abort_if_reached() { \
+        assert(false);             \
+    }                              \
+}
```
Comment (vpodzime) [inline libutils/condition_macros.h]: Extra `}` here?

Comment (olehermanse) [reply libutils/condition_macros.h]: Fixed, thanks!

File: libutils/condition_macros.h (hunk 1 above)
Comment (olehermanse) [reply libutils/condition_macros.h]: Thanks, fixed!

File: libutils/condition_macros.h [hunk 3]
```diff
@@ -17,4 +26,12 @@
 }
 #endif
 
+// Similar to assert_or_return, except you put it inside the if
+// body which handles the error in release builds
+#ifndef debug_abort_if_reached
+#define debug_abort_if_reached() { \
+    assert(false);                 \
```
Comment (karlhto) [inline libutils/condition_macros.h]: I guess the only problem here is that you aren't told exactly what condition is wrong when it fails. Adding a parameter to the macro would kind of remove the point of having it in the first place, however.

Comment (olehermanse) [reply libutils/condition_macros.h]: It should be placed inside an if, and you get a line number. It should be pretty clear.

Comment (olehermanse) [reply libutils/condition_macros.h]: I hope you get a line number. Good point, when I statrt using it I can do some testing and improve it if necessary.

**Comment by olehermanse** [comment]
Cherry-picked in https://github.com/cfengine/core/pull/3825

**Pull request merged**